Github Workflow Release

jeffsuperglide opened the pull request:
- release.yml
  - ubuntu and macos
  - on.push.tags pattern = "[0-9]+-[A-Z]+-[0-9]+"
  - creates release with `github.ref_name`
  - assest naming `javaHeclib_${{ matrix.os }}_${{ github.ref_name }}.tar.gz`
  - release body takes markdown from RELEASE.md
- RELEASE.md file

TODO: Add Windows build as additional job

ktarbet commented:
@jeffsuperglide  Thanks for this work!

Would you please describe how the RELEASE.md is used?

jeffsuperglide commented:
> @jeffsuperglide Thanks for this work!
> 
> Would you please describe how the RELEASE.md is used?

Add markdown to `RELEASE.md` and it will render in the body of the release.  Each change/fix/update can be noted in this file or appended/prepended.  The thought is something like the following:

### Release 7-IP

- Fixed issue #???
- Added support for...

### Release 7-IO

***This release fixes all the stuff Jeff did*** 😆

### Release 7-IN

- Dropped support for Windows 👎 
- etc...

I just noticed that the version is without `-#` at the end.  The tag within the workflow could change or start adding a `patch`; assuming the versioning is major-minor-patch.

ktarbet commented:
We need to discuss possibly changing  where the artifacts will live.   Right now multiple projects are using  a nexus instance to retrieve hec-dss binaries as dependencies (javaHeclib.dll, and heclib_c.lib, and heclib_f.lib).   

I think we want to have a single source of binaries. Would the binaries be copied from github to nexus to keep to a single source, but with copies?

jeffsuperglide commented:
Some times on leave I would look at the Makefiles and try to make them less OS dependent (i.e., not using `shell`, `find`, `ls`, `uname`, etc) but couldn't get macOS to build the last step (javaHeclib).  No matter what I did archive didn't like some of the objects.  My makefiles had a lot less code and could determine OS and arch.  I eventually tossed it b/c it wouldn't have built on Team City anyway.

This PR should produce the exact same .so and .dylib on Team City b/c it's the same Makefiles.

jeffsuperglide commented:
I could see what comes with the GitHub runner images (https://github.com/actions/runner-images/tree/main/images) but not what is setup on Team City.  Made some assumptions based on Makefiles.

MikeNeilson commented on .github/workflows/builds.yml:
```diff
@@ -7,26 +7,35 @@ env:
+      matrix:
+        os: [ubuntu-20.04, macos-11]
+    runs-on: ${{ matrix.os }}
+    if: startsWith(matrix.os, 'ubuntu')
```
don't we want the tests to run on the macos runner?

jeffsuperglide replied:
> don't we want the tests to run on the macos runner?

Of course.  The existing macOS workflow does not test the macOS runner.  Thinking to add it but testing is a pain.  The `startsWith` or `contains` does not like any iteration of ${{ matrix.os }}, '${{ matrix.os }}', or just matrix.os.  VS Code intelliSense doesn't have an issue but GitHub is angry with them.

MikeNeilson replied:
okay, no I had something similarish,  https://github.com/USACE/cwms-radar-api/blob/fc8a55a8970ebfad4a414e9dd4e484a7882a26c0/.github/workflows/build.yml#L111 but that doesn't really give you an example of the matrix there.
